### PR TITLE
spatial transformations

### PR DESCRIPTION
**Pull request by PascalSenn**
Adds reprojection features to spatial types

### CODE REVIEW

File: src/HotChocolate/Spatial/src/Types/Transformation/GeometryProjector.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        private ICoordinateTransformation _transformation;
+
+        public GeometryProjector(ICoordinateTransformation transformation)
+        {
+            _transformation = transformation;
+        }
+
+        public void Reproject(Geometry geometry, int targetSrid)
+        {
+            Parallel.ForEach(geometry.Coordinates, AdjustCoordinate);
```
Comment (steveoh) [inline src/HotChocolate/Spatial/src/Types/Transformation/GeometryProjector.cs:19]: oooh fancy, nice touch.

**Comment by steveoh** [comment]
This looks pretty nice.

**Comment by steveoh** [comment]
It would be nice to add these two defaults from NTS that will be the most popular/common spatial references that folks will be using https://github.com/NetTopologySuite/ProjNet4GeoAPI/blob/6a6ee1c9c1eb27877c43e0a58ae280f7510ef444/src/ProjNet/CoordinateSystemServices.cs#L197-L198.

I would propose to add an extension method to set `WGS84` or `WebMercator`. The WKT string that you have will be great for everything else that NTS can support.

**Comment by PascalSenn** [comment]
@steveoh  added webmercator and WGS84 as extension methods

File: src/HotChocolate/Spatial/src/Types/Transformation/GeometryTransformerInterceptor.cs [hunk 2]
```diff
@@ -23,7 +23,8 @@ internal class GeometryTransformerInterceptor : TypeInterceptor
             IDictionary<string, object?> contextData)
         {
             ISpatialConvention convention = completionContext.GetSpatialConvention();
-            if (convention.TransformerFactory.HasCoordinateSystems())
+            if (convention.TransformerFactory.HasCoordinateSystems() &&
+                convention.DefaultSrid is not 0 and not -1)
```
Comment (steveoh) [inline src/HotChocolate/Spatial/src/Types/Transformation/GeometryTransformerInterceptor.cs:27]:
```suggestion
                convention.DefaultSrid > 0)
```

File: src/HotChocolate/Core/src/Abstractions/ErrorCodes.cs [hunk 3]
```diff
@@ -115,6 +115,18 @@ public static class Stitching
...
+            /// The coordinate reference system is not supported by this server
+            /// </summary>
+            public const string UnknowCrs = "HC0029";
+
+            /// <summary>
+            /// Coordinates with M values cannot be reprojected
+            /// </summary>
+            public const string CoordinateMNotSupported = "HC0030";
+        }
         public static class Data
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Abstractions/ErrorCodes.cs:130]: line

File: src/HotChocolate/Spatial/demo/Spatial.Demo/Startup.cs [hunk 4]
```diff
@@ -22,7 +22,15 @@ public void ConfigureServices(IServiceCollection services)
...
+                    .RegisterCoordinateSystem(4326,
+                        "GEOGCS[\"WGS 84\",DATUM[\"WGS_1984\",SPHEROID[\"WGS 84\",6378137,298.257223563,AUTHORITY[\"EPSG\",\"7030\"]],AUTHORITY[\"EPSG\",\"6326\"]],PRIMEM[\"Greenwich\",0,AUTHORITY[\"EPSG\",\"8901\"]],UNIT[\"degree\",0.0174532925199433,AUTHORITY[\"EPSG\",\"9122\"]],AUTHORITY[\"EPSG\",\"4326\"]]")
+                    .RegisterCoordinateSystem(26918,
+                        "PROJCS[\"NAD83 \\/ UTM zone 18N\",GEOGCS[\"NAD83\",DATUM[\"North_American_Datum_1983\",SPHEROID[\"GRS 1980\",6378137,298.257222101,AUTHORITY[\"EPSG\",\"7019\"]],TOWGS84[0,0,0,0,0,0,0],AUTHORITY[\"EPSG\",\"6269\"]],PRIMEM[\"Greenwich\",0,AUTHORITY[\"EPSG\",\"8901\"]],UNIT[\"degree\",0.0174532925199433,AUTHORITY[\"EPSG\",\"9122\"]],AUTHORITY[\"EPSG\",\"4269\"]],PROJECTION[\"Transverse_Mercator\"],PARAMETER[\"latitude_of_origin\",0],PARAMETER[\"central_meridian\",-75],PARAMETER[\"scale_factor\",0.9996],PARAMETER[\"false_easting\",500000],PARAMETER[\"false_northing\",0],UNIT[\"metre\",1,AUTHORITY[\"EPSG\",\"9001\"]],AXIS[\"Easting\",EAST],AXIS[\"Northing\",NORTH],AUTHORITY[\"EPSG\",\"26918\"]]"
+                    ))
```
Comment (michaelstaib) [inline src/HotChocolate/Spatial/demo/Spatial.Demo/Startup.cs:33]: needs to be on 32

File: src/HotChocolate/Spatial/src/Types/HotChocolate.Types.Spatial.csproj [hunk 5]
```diff
@@ -14,6 +14,7 @@
 
   <ItemGroup>
     <PackageReference Include="NetTopologySuite" Version="2.0.0" />
+    <PackageReference Include="ProjNET" Version="2.0.0" />
```
Comment (michaelstaib) [inline src/HotChocolate/Spatial/src/Types/HotChocolate.Types.Spatial.csproj:17]: what is projNET

File: src/HotChocolate/Spatial/src/Types/Transformation/GeometryTransformer.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                double y;
+                double z;
+
+                switch (coordinate)
+                {
+                    case CoordinateZM:
+                    case CoordinateM:
+                        hasErrors = true;
+                        break;
+                    case CoordinateZ:
```
Comment (michaelstaib) [inline src/HotChocolate/Spatial/src/Types/Transformation/GeometryTransformer.cs:37]: line

File: src/HotChocolate/Spatial/src/Types/Transformation/GeometryTransformer.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                        x = coordinate.X;
+                        y = coordinate.Y;
+                        z = coordinate.Z;
+                        _transformation.MathTransform.Transform(ref x, ref y, ref z);
+                        coordinate.X = x;
+                        coordinate.Y = y;
+                        coordinate.Z = z;
+                        break;
+                    default:
```
Comment (michaelstaib) [inline src/HotChocolate/Spatial/src/Types/Transformation/GeometryTransformer.cs:46]: line

File: src/HotChocolate/Spatial/src/Types/HotChocolate.Types.Spatial.csproj (hunk 5 above)
Comment (steveoh) [reply src/HotChocolate/Spatial/src/Types/HotChocolate.Types.Spatial.csproj:17]: It's the nts projection project.

File: src/HotChocolate/Core/src/Abstractions/ErrorCodes.cs (hunk 3 above)
Comment (PascalSenn) [reply src/HotChocolate/Core/src/Abstractions/ErrorCodes.cs:130]: done

**Comment by steveoh** [comment]
@michaelstaib do you have any comments besides code style?

**Comment by michaelstaib** [comment]
Is this done?

File: src/HotChocolate/Spatial/demo/Spatial.Demo/Startup.cs [hunk 8]
```diff
@@ -22,7 +23,14 @@ public void ConfigureServices(IServiceCollection services)
                 .AddGraphQLServer()
                 .AddFiltering()
                 .AddProjections()
-                .AddSpatialTypes()
+                .AddSpatialTypes(x => x
+                    .DefaultSrid(26912)
+                    .AddWebMercator()
+                    .AddWGS84()
+                    .AddCoordinateSystemFromString(26912,
+                        "PROJCS[\"NAD83 \\/ UTM zone 12N\",GEOGCS[\"NAD83\",DATUM[\"North_American_Datum_1983\",SPHEROID[\"GRS 1980\",6378137,298.257222101,AUTHORITY[\"EPSG\",\"7019\"]],TOWGS84[0,0,0,0,0,0,0],AUTHORITY[\"EPSG\",\"6269\"]],PRIMEM[\"Greenwich\",0,AUTHORITY[\"EPSG\",\"8901\"]],UNIT[\"degree\",0.0174532925199433,AUTHORITY[\"EPSG\",\"9122\"]],AUTHORITY[\"EPSG\",\"4269\"]],PROJECTION[\"Transverse_Mercator\"],PARAMETER[\"latitude_of_origin\",0],PARAMETER[\"central_meridian\",-111],PARAMETER[\"scale_factor\",0.9996],PARAMETER[\"false_easting\",500000],PARAMETER[\"false_northing\",0],UNIT[\"metre\",1,AUTHORITY[\"EPSG\",\"9001\"]],AXIS[\"Easting\",EAST],AXIS[\"Northing\",NORTH],AUTHORITY[\"EPSG\",\"26912\"]]")
```
Comment (michaelstaib) [inline src/HotChocolate/Spatial/demo/Spatial.Demo/Startup.cs:31]: line length

File: src/HotChocolate/Spatial/src/Types/Configuration/ISpatialConvention.cs [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace HotChocolate.Types.Spatial.Configuration
+{
+    /// <summary>
+    /// A conventions that configures the behaviour of spatial types
+    /// </summary>
+    public interface ISpatialConvention : IConvention
+    {
+        /// <summary>
+        /// The default SRID/CRS of the schema. All incoming queries will be translated to this SRID
```
Comment (michaelstaib) [inline src/HotChocolate/Spatial/src/Types/Configuration/ISpatialConvention.cs:12]: line length?

File: src/HotChocolate/Spatial/src/Types/Configuration/ISpatialConventionDescriptor.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace HotChocolate.Types.Spatial.Configuration
+{
+    /// <summary>
+    /// The descriptor of a <see cref="SpatialConvention"/>
+    /// </summary>
+    public interface ISpatialConventionDescriptor
+    {
+        /// <summary>
+        /// The default SRID/CRS of the schema. All incoming queries will be translated to this SRID
```
Comment (michaelstaib) [inline src/HotChocolate/Spatial/src/Types/Configuration/ISpatialConventionDescriptor.cs:11]: line length?

File: src/HotChocolate/Spatial/src/Types/Transformation/GeometryTransformer.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            _transformation = transformation;
+        }
+
+        /// <inheritdoc />
+        public void TransformInPlace(Geometry geometry, int targetSrid)
+        {
+            bool hasErrors = false;
+
+            // This function transforms the coordinate in place. The instance of the coordinate
```
Comment (michaelstaib) [inline src/HotChocolate/Spatial/src/Types/Transformation/GeometryTransformer.cs:23]: line length?

File: src/HotChocolate/Spatial/demo/Spatial.Demo/Startup.cs (hunk 8 above)
Comment (PascalSenn) [reply src/HotChocolate/Spatial/demo/Spatial.Demo/Startup.cs:31]: Should i split the string?!

File: src/HotChocolate/Spatial/src/Types/Configuration/ISpatialConvention.cs (hunk 9 above)
Comment (PascalSenn) [reply src/HotChocolate/Spatial/src/Types/Configuration/ISpatialConvention.cs:12]: Living on the edge
image

File: src/HotChocolate/Spatial/src/Types/Configuration/ISpatialConventionDescriptor.cs (hunk 10 above)
Comment (PascalSenn) [reply src/HotChocolate/Spatial/src/Types/Configuration/ISpatialConventionDescriptor.cs:11]: image

**Comment by steveoh** [comment]
@michaelstaib do you have any review comments besides code style?

**Pull request merged**